components/kyma-adapter/internal/handler/handler.go

```diff
@@ -1,16 +1,187 @@
+		respondWithSuccess(ctx, w, configPendingState, fmt.Sprintf("Configuration is missing. Responding with %q state...", configPendingState))
+		return
+	}
+
+	appID := reqBody.GetApplicationID()
+	rtmID := reqBody.GetRuntimeID()
+
+	log.C(ctx).Infof("Getting application bundles for app with id %q and tenant %q", appID, tenantID)
+	bundles, err := a.DirectorGqlClient.GetApplicationBundles(ctx, appID, tenantID)
+	if err != nil {
+		respondWithError(ctx, w, http.StatusBadRequest, createErrorState, errors.Wrapf(err, "while getting application bundles"))
+		return
+	}
+
+	if len(bundles) == 0 {
+		respondWithSuccess(ctx, w, readyState, fmt.Sprintf("There are no bundles for application with ID %q", appID))
+		return
+	}
+
+	instanceAuthExist := false
+	for _, instanceAuth := range bundles[0].InstanceAuths {
+		if *instanceAuth.RuntimeID == rtmID {
+			instanceAuthExist = true
+			break
+		}
+	}
+
+	if !instanceAuthExist && operation == unassignOperation {
+		respondWithSuccess(ctx, w, readyState, fmt.Sprintf("There are no bundle instance auths for deletion for runtime with ID %q and application with id %q", rtmID, appID))
+		return
+	}
+
+	creds := credentials.NewCredentials(reqBody)
+	modifyFunc := a.determineAuthModifyFunc(w, instanceAuthExist, operation)
+	for _, bundle := range bundles {
+		input := buildInstanceAuthInput(instanceAuthExist, operation, bundle, rtmID, creds)
+		if err = modifyFunc(ctx, tenantID, input); err != nil {
+			return
+		}
+	}
+
+	respondWithSuccess(ctx, w, readyState, "Successfully processed Kyma integration.")
+}
+
+func (a AdapterHandler) determineAuthModifyFunc(w http.ResponseWriter, authExists bool, operation string) modifyBundleInstanceAuthFunc {
+	if operation == assignOperation {
+		if authExists {
+			// Update func
+			return func(ctx context.Context, tenant string, input gqlclient.BundleInstanceAuthInput) (err error) {
+				if err = a.DirectorGqlClient.UpdateBundleInstanceAuth(ctx, tenant, input); err != nil {
+					respondWithError(ctx, w, http.StatusBadRequest, createErrorState, err)
+				}
+				return err
+			}
+		} else {
+			// Create func
+			return func(ctx context.Context, tenant string, input gqlclient.BundleInstanceAuthInput) (err error) {
+				if err = a.DirectorGqlClient.CreateBundleInstanceAuth(ctx, tenant, input); err != nil {
+					respondWithError(ctx, w, http.StatusBadRequest, createErrorState, err)
+				}
+				return err
+			}
+		}
+	} else {
+		// Delete func
+		return func(ctx context.Context, tenant string, input gqlclient.BundleInstanceAuthInput) (err error) {
+			if err = a.DirectorGqlClient.DeleteBundleInstanceAuth(ctx, tenant, input); err != nil {
+				respondWithError(ctx, w, http.StatusBadRequest, deleteErrorState, err)
+			}
+			return err
+		}
+	}
+}
+
+func buildInstanceAuthInput(authExists bool, operation string, bundle *graphql.BundleExt, rtmID string, credentials credentials.Credentials) gqlclient.BundleInstanceAuthInput {
+	if operation == assignOperation {
```

Comment:
Can we reuce the nesting as it gets harder to follow the code. We could change it to
`if operation == assignOperation && authExists {update}
else if operation == assignOperation && !authExists {create}
else {delete}`